src/hmm/topology.cc

```diff
@@ -239,148 +118,178 @@ void Topology::Check() {
-          "have at least one emitting state).";
-    if (!entries_[i][num_states-1].transitions.empty())
-      KALDI_ERR << "Topology::Check(), last state must have no transitions.";
-    // not sure how necessary this next stipulation is.
-    if (entries_[i][num_states-1].forward_pdf_class != kNoPdf)
-      KALDI_ERR << "Topology::Check(), last state must not be emitting.";
-
-    std::vector<bool> has_trans_in(num_states, false);
-    std::vector<int32> seen_pdf_classes;
+      KALDI_ERR << "Topology::Check(), cannot only have one state (must have a "
+                << "final state and a start state).";
+  }
 
-    for (int32 j = 0; j < num_states; j++) {  // j is the state-id.
-      BaseFloat tot_prob = 0.0;
-      if (entries_[i][j].forward_pdf_class != kNoPdf) {
-        seen_pdf_classes.push_back(entries_[i][j].forward_pdf_class);
-        seen_pdf_classes.push_back(entries_[i][j].self_loop_pdf_class);
-      }
-      std::set<int32> seen_transition;
-      for (int32 k = 0;
-           static_cast<size_t>(k) < entries_[i][j].transitions.size();
-           k++) {
-        tot_prob += entries_[i][j].transitions[k].second;
-        if (entries_[i][j].transitions[k].second <= 0.0)
-          KALDI_ERR << "Topology::Check(), negative or zero transition prob.";
-        int32 dst_state = entries_[i][j].transitions[k].first;
-        // The commented code in the next few lines disallows a completely
-        // skippable phone, as this would cause to stop working some mechanisms
-        // that are being built, which enable the creation of phone-level lattices
-        // and rescoring these with a different lexicon and LM.
-        if (dst_state == num_states-1 // && j != 0
-            && entries_[i][j].forward_pdf_class == kNoPdf)
-          KALDI_ERR << "We do not allow any state to be "
-              "nonemitting and have a transition to the final-state (this would "
-              "stop the SplitToPhones function from identifying the last state "
-              "of a phone.";
-        if (dst_state < 0 || dst_state >= num_states)
-          KALDI_ERR << "Topology::Check(), invalid dest state " << (dst_state);
-        if (seen_transition.count(dst_state) != 0)
-          KALDI_ERR << "Topology::Check(), duplicate transition found.";
-        if (dst_state == k) {  // self_loop...
-          KALDI_ASSERT(entries_[i][j].self_loop_pdf_class != kNoPdf &&
-                       "Nonemitting states cannot have self-loops.");
-        }
-        seen_transition.insert(dst_state);
-        has_trans_in[dst_state] = true;
+  bool has_final_state = false;
+  for (auto const& entry: entries_) {
+    for (fst::StateIterator<fst::StdVectorFst> state_iter(entry);
+         !state_iter.Done(); state_iter.Next()) {
+      StateId state_id = state_iter.Value();
+      if (entry.Final(state_id) != Weight::Zero()) {
+        has_final_state = true;
+
+        // TODO(galv): Figure out whether this is really
```

Comment:
It is not necessary for final states to have no arcs leaving them  (although there must not be transitions *into* the start state).  Any statement otherwise should have been removed I think, maybe it was in an early draft.